### ISSUE TITLE
Annotations break color cycles

### ISSUE THREAD

**Issue by jbednar** [ioam/holoviews#2593]
Usually, calling .overlay() on a holomap will give plots that follow a color cycle: 

```
measles_by_state = ds.to(hv.Curve, 'Year', 'measles')
```

![image](https://user-images.githubusercontent.com/1695496/39082137-e37b51d0-4513-11e8-8fa5-493f71741e8c.png)

However, if the underlying elements are annotated (and thus actually Overlays and not Elements), the colors do not cycle:

```
measles_by_state = ds.to(hv.Curve, 'Year', 'measles') * hv.VLine(1963).options(color='black')
```

!![image](https://user-images.githubusercontent.com/1695496/39082130-c3c89faa-4513-11e8-8c8f-d3354312f109.png)

It seems to me that if we have a concept of Annotation, we should make annotating things change their behavior as little as possible.

**Comment by philippjfr** [comment]
I'd very much recommend against creating NdOverlays of Overlays in almost all scenarios, precisely for reasons like this. There is no easy fix here and I'm not sure there will ever be. Only thing I could think of is to dynamically reorganize the datastructure before it is plotted a bit like ``collate``.

**Comment by jbednar** [comment]
As a general rule, I would argue strongly that Annotations should be as transparently supported as possible.  I realize that once it's been applied, overlaying an Annotation is simply like any other Overlay to HoloViews.  But to a user, an Annotation is something tied very closely to a plot, like a legend, and any time that having the Annotation disrupts some functionality or reduces what you can do with the plot is confusing and frustrating. 

Looking at the bigger picture, one of the strongest arguments in favor of using HoloViews compared to a regular plotting program is that HV lets you capture your understanding of a dataset in a way that propagates as that dataset is transformed and selected.  Annotations are a big part of that -- "something happened *here* in the data; see?", and the earlier we can tag the data with that information without disrupting further operations, the better.  

So if you have a HoloViews object `o` that you put an annotation `a` on (as with the VLine overlaid on a HoloMap in this example), `a` should, where possible, (1) not cause errors when `o*a` is used in a way that's valid for `o` alone, and (2) should result in the same as doing that thing to `o` alone, then overlaying `a` again.  

Making that work does seem like it would require something like `collate`, but I think it's the right thing to do if we want to support users in capturing their understanding as they achieve it.

**Comment by philippjfr** [comment]
I'll just briefly interject that this has nothing to do with annotations, you simply structured the datastructure in a way that is explicitly not recommended and you'd face the same issues with any non-annotation element too. If you overlaid the annotation last on top of the ``NdOverlay``, as is recommended, you wouldn't have this problem. If someone wants to tackle this special type of collation they can be my guest but it's bound to be extremely hairy code and since we explicitly recommend against this it's unlikely to get prioritized.

**Comment by jlstevens** [comment]
I agree with what Philipp just said and if we rely on ``collate`` too much, things *will* break in hard to understand ways. Collate can fix things up a bit but I wouldn't want to rely on it to fix all possible nestings of things...

**Comment by jbednar** [comment]
I agree that the recommended data structure is a well defined, clear way to do something to get the final result desired. I also understand that from the point of view of HoloViews' internal code, Annotations are no different from any other Overlay in principle.  But I'm making what I think is an important point about usability for the types of things we think and say that HoloViews is meant to be useful for.

I should probably explain the actual sequence of events that led to the above issue.  Chronologically, the ordering was:

1. I made a HoloMap of measles incidence by State:
![image](https://user-images.githubusercontent.com/1695496/39140621-c183d83a-46ea-11e8-95ad-ec5187380958.png)

2. I annotated it with a vertical line indicating when the vaccine was introduced in the USA.
![image](https://user-images.githubusercontent.com/1695496/39140632-cadaa012-46ea-11e8-928b-76e2ec962826.png)

3. After playing around with the widgets, I decided to lay out a few of them left to right, which preserves the annotation very nicely:
![image](https://user-images.githubusercontent.com/1695496/39140650-da3e93ba-46ea-11e8-9f25-a203ba140372.png)

4. Then I wanted to make an overlay instead, which works, but doesn't color cycle, so it's not useful:
![image](https://user-images.githubusercontent.com/1695496/39140699-03ae9d4e-46eb-11e8-85b8-25a598dd2c9b.png)

Because step 4 didn't work, I had to go back and tell the entire story in a different way, inserting an annotation onto every plot where I wanted it.  Sure, that works, but it undercuts one of the main reasons to use HoloViews, which is to capture your understanding of a dataset, declaratively, at the earliest possible stage.  Here, I now have to keep my understanding (expressed as the 1963 annotation) in my head, reapplying it whenever I use this data, instead of being able to express it when I first wanted to, and having it travel along with the data.  That's the sense in which an Annotation is different from data that can also form an Overlay -- it's metadata, and should propagate.

Sure, it can get hairy actually doing such propagation.  But what I'm arguing here is that it's desirable, so that people are actively encouraged to add annotations and other declarations as early as possible, having them apply to subsequent operations programmatically rather than because I as a programmer kept them in my head and applied them explicitly each time.

**Comment by philippjfr** [comment]
I'm going to come at this from a slightly different angle because I've now thought about this a bit more. While I did say that NdOverlays of Overlays are not recommended they are technically valid and quite distinct from Overlays of NdOverlays. Personally I don't see any immediate usecases for them but we explicitly did not disallow them so it is possible that someone might be using them. Magically restructuring therefore has the potential to completely change the semantics of the declared datastructure. 

Take this example where I create two datasets which might contain some observations of x/y grouped by category and group taken from different sources 'A' and 'B':

```python
ds = hv.Dataset({('x', 'y'): np.random.rand(100, 2), 'group': np.random.randint(0, 5, 100), 'category': np.random.randint(0, 3, 100)}, ['group', 'category', 'x'], ['y'], label='A')
ds2 = hv.Dataset({('x', 'y'): np.random.rand(100, 2), 'group': np.random.randint(0, 5, 100), 'category': np.random.randint(0, 3, 100)}, ['group', 'category', 'x'], ['y'], label='B')
```

Now I want to view both of these datasets together, so I'd create an Overlay of NdOverlays:

```python
(ds.to(hv.Scatter, 'x', 'y').overlay('category') * ds2.to(hv.Scatter, 'x', 'y').overlay('category'))
```

<img width="903" alt="screen shot 2018-04-23 at 6 28 42 pm" src="https://user-images.githubusercontent.com/1550771/39143039-795061e8-4724-11e8-9d08-14f6195790f6.png">

At least in theory I might instead want to not compare between the categories but instead compare between the two datasets. I now have two options, I can ignore the 'category' by declaring:

```python
(ds.to(hv.Scatter, 'x', 'y', 'group') * ds2.to(hv.Scatter, 'x', 'y', 'group))
```

But maybe I've already built a whole narrative around comparing the two so I've come up with something like this:

```python
scatter_comparison = ds.to(hv.Scatter, 'x', 'y') * ds2.to(hv.Scatter, 'x', 'y') *  hv.Ellipse(0.5, 0.5, 0.2)
```

And have now decided that actually I don't care about the 'category' variable so I do this:

```python
scatter_comparison.overlay('category')
```

<img width="913" alt="screen shot 2018-04-23 at 6 38 33 pm" src="https://user-images.githubusercontent.com/1550771/39143333-8f501122-4725-11e8-91b6-4419acfdacda.png">

This is a valid thing to do even though I wouldn't necessarily recommend it. So I have serious reservations about reordering the users data and thereby making an assumption what the user is trying to achieve. If we do not explicitly disallow a particular organization then we should not then break the semantics, which that organization implies. In the case of an annotation that's somewhat annoying because those semantics do not necessarily apply there, but imo that's a price we have to pay for internal consistency.

**Comment by jbednar** [comment]
That's much more compelling, thanks.  That argument would suggest providing a utility and an example of using it to make the color cycling work above, so that when this situation comes up, there is a recommended way to reorganize the object, not just a recommendation to start over and build your object up a different way.
